node-graph/gstd/src/raster.rs

```diff
@@ -130,17 +130,86 @@ pub struct BlendImageNode<Second, MapFn> {
 	map_fn: MapFn,
 }
 
+fn sample_from_image(image: &Image, u: f64, v: f64) -> Color {
+	let x = ((u * image.width as f64) as u32).clamp(0, image.width);
+	let y = ((v * image.height as f64) as u32).clamp(0, image.height);
+
+	image.data[(y * image.width + x) as usize]
+}
+
+struct AxisAlignedBbox {
+	start: DVec2,
+	end: DVec2,
+}
+
+struct Bbox {
+	top_left: DVec2,
+	top_right: DVec2,
+	bottom_left: DVec2,
+	bottom_right: DVec2,
+}
+
+impl Bbox {
+	fn axis_aligned_bbox(&self) -> AxisAlignedBbox {
+		let start_x = self.top_left.x.min(self.top_right.x).min(self.bottom_left.x).min(self.bottom_right.x);
+		let start_y = self.top_left.y.min(self.top_right.y).min(self.bottom_left.y).min(self.bottom_right.y);
+		let end_x = self.top_left.x.max(self.top_right.x).max(self.bottom_left.x).max(self.bottom_right.x);
+		let end_y = self.top_left.y.max(self.top_right.y).max(self.bottom_left.y).max(self.bottom_right.y);
+
+		AxisAlignedBbox {
+			start: DVec2::new(start_x, start_y),
+			end: DVec2::new(end_x, end_y),
+		}
+	}
+}
+
+fn compute_transformed_bounding_box(image_frame: &ImageFrame) -> Bbox {
+	let top_left = DVec2::new(0.0, 0.0);
+	let top_right = DVec2::new(image_frame.image.width as f64, 0.0);
+	let bottom_left = DVec2::new(0.0, image_frame.image.height as f64);
+	let bottom_right = DVec2::new(image_frame.image.width as f64, image_frame.image.height as f64);
+	let transform = |p| image_frame.transform.transform_point2(p);
+
+	Bbox {
+		top_left: transform(top_left),
+		top_right: transform(top_right),
+		bottom_left: transform(bottom_left),
+		bottom_right: transform(bottom_right),
+	}
+}
+
 // TODO: Implement proper blending
 #[node_macro::node_fn(BlendImageNode)]
 fn blend_image<MapFn>(image: ImageFrame, second: ImageFrame, map_fn: &'any_input MapFn) -> ImageFrame
 where
 	MapFn: for<'any_input> Node<'any_input, (Color, Color), Output = Color> + 'input,
 {
-	let mut image = image;
-	for (pixel, sec_pixel) in &mut image.image.data.iter_mut().zip(second.image.data.iter()) {
-		*pixel = map_fn.eval((*pixel, *sec_pixel));
+	let (mut dst_image, src_image) = (image, second);
+	let (src_width, src_height) = (src_image.image.width, src_image.image.height);
+	let (dst_width, dst_height) = (dst_image.image.width, dst_image.image.height);
+	let aabb = compute_transformed_bounding_box(&src_image).axis_aligned_bbox();
+	let (start_x, end_x) = ((aabb.start.x as i64).max(0) as u32, (aabb.end.x as i64).min(dst_width as i64) as u32);
+	let (start_y, end_y) = ((aabb.start.y as i64).max(0) as u32, (aabb.end.y as i64).min(dst_height as i64) as u32);
```

Comment:
I would assume that you would have to do something like this: (assuming flipped src and dest)
```rust
// transform that converts a point from background space to the new_image space
let bg_to_fg = fg.transform.inverse() * bg.transform;

for y in start_y..end_y {
    for x in  start_x..end_x {
        let point = DVec2::new(x as f64, y as f64);
        // bounds check
        let bg_pixel = &mut bg_image.image.get_mut(x,y); // add a functions on image to make this more convenient
        let fg_pixel = fg_image.sample(bg_to_fg.transform_point_2(point))
        *bg_pixel = map_fn.eval((*dst_pixel, src_pixel));
    }
}